design/one-pager-cli-diff.md

```diff
@@ -0,0 +1,187 @@
+# `crossplane beta diff` command
+
+* Owner: Jonathan Ogilvie (@jcogilvie)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+The Crossplane community has [long desired](https://github.com/crossplane/crossplane/issues/1805) a mechanism to 
+understand the would-be surface area of proposed changes to Crossplane resources. Platform administrators and 
+application teams would like the ability to review changes before they are applied, especially in GitOps workflows. 
+This requirement is particularly critical when working with complex compositions that result in multiple managed 
+resources being created or updated.
+
+Currently, users lack visibility into how changes to Composite Resources (XRs) or their configurations translate to 
+changes in the underlying Managed Resources (MRs) and ultimately in the external systems. This makes it challenging to 
+understand the impact of changes, troubleshoot issues, and confidently apply updates to production environments.
+
+## Goals
+
+The `crossplane beta diff` command aims to:
+
+1. Provide a clear, familiar way to preview changes that would be made to external resources before they are applied
```

Comment:
Are there any kinds of change that are out of scope to preview? (e.g. provider upgrades)

Who do we expect to run this CLI? It seems as if the entity running the CLI will need almost all of the read permissions Crossplane has, in order to compute the diff. e.g. They'll need to be able to read:

* The claim
* The XR
* The composition (or compositions, for selection)
* All composed resources (including nested XRs)
* Any extra resources

The average app developer working with claims may not have access to all these things, depending on how the org chooses to setup their RBAC.

Reply:
MVP is diffs of XRs/MRs for a given composition already found in the cluster.  Composition upgrades or revisions are probably vNext.  Provider upgrades are currently out of scope, but since the unit of change there is mostly the schemas, we could probably wrap something around `validate` as an additional tool.  

The primary persona I have in mind is the CI runner.  This diff gets posted on your gitops PR.  Other personas are certainly worth considering, and we could support tighter-scoped permissions by allowing the user to provide extra objects on the command line or as files, but I think that's probably vNext.